python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+        for priority in priorities:
+            # iterate through the tasks and make sure we aren't waiting on the
+            # completion of any upstream tasks:
+            for ti, task in enumerate(self._pending_tasks_by_priority[priority]):
+                awaiting_upstream_task_completion = False
+                for us_task_id in self._upstream_task_map.get(task.uid, []):
+                    if us_task_id in self._tasks_by_id:
+                        # if the task is still in the tasks list then we're still awaiting
+                        # completion of it!
+                        awaiting_upstream_task_completion = True
+                        break
+                if awaiting_upstream_task_completion:
+                    continue
+
+                # ok, we've found the next task to process:
+                task_to_process = task
+                task_index = ti
+                break
+
+            if task_to_process:
+                # no need to look any further!
+                break
+
+        if not task_to_process:
+            # nothing to do!
+            return False
+
+        # we need a thread to do the work with:
+        thread = self._get_worker_thread()
+        if not thread:
+            # looks like we can't do anything!
+            return False
+
+        self._log("Starting task %r" % task_to_process)
+
+        # ok, we have a thread so lets move the task from the priority queue to the running list:
+        self._pending_tasks_by_priority[priority] = (self._pending_tasks_by_priority[priority][:task_index] 
+                                                     + self._pending_tasks_by_priority[priority][task_index+1:])
+        if not self._pending_tasks_by_priority[priority]:
+            # no more tasks with this priority so also clean up the list
+            del self._pending_tasks_by_priority[priority]
+        self._running_tasks[task_to_process.uid] = (task_to_process, thread)
+
+        num_tasks_left = 0
+        for pending_tasks in self._pending_tasks_by_priority.values():
+            num_tasks_left += len(pending_tasks)
+        self._log(" > Currently running tasks: '%s' - %d left in queue" % (self._running_tasks.keys(), num_tasks_left))
+
+        # and run the task
+        thread.run_task(task_to_process)
+
+        return True
+
+    def _on_task_completed(self, task, result):
+        """
+        Slot triggered when a task is completed.  This processes the result and emits the 
+        task_completed signal if needed.
+
+        :param task:    The task that completed
+        :param result:  The task result
+        """
+        try:
+            # check that we should process this result:
+            if task.uid in self._running_tasks:
+                self._log("Task %r - completed" % (task))
+
+                # if we have dependent tasks then update them:
+                for ds_task_id in self._downstream_task_map.get(task.uid) or []:
```

Comment:
A quick comment why get could return None would be appreciated. From my understanding, it's because it's possible on of the task that depends on you could have been removed for some reason before and therefore get would return None. Which begs the question, why not .get(task.uid, [])?